File: packages/react-native-autoplay/ios/templates/MessageTemplate.swift
//
//  MessageTemplate.swift
//  Pods
//
//  Created by Samuel Brucksch on 17.10.25.
//

import CarPlay

class MessageTemplate: AutoPlayTemplate {
    var config: MessageTemplateConfig

    init(config: MessageTemplateConfig) {
        self.config = config

        let template = CPAlertTemplate(
            titleVariants: [Parser.parseText(text: config.message)!],
            actions: Parser.parseAlertActions(alertActions: config.actions),
            id: config.id
        )

        super.init(
            template: template,
            header: config.headerActions
        )
    }

    override func onWillAppear(animated: Bool) {
        config.onWillAppear?(animated)
    }

    override func onDidAppear(animated: Bool) {
        config.onDidAppear?(animated)
    }

    override func onWillDisappear(animated: Bool) {
        config.onWillDisappear?(animated)
    }

    override func onDidDisappear(animated: Bool) {
        config.onDidDisappear?(animated)
    }

    override func onPopped() {
        config.onPopped?()
    }
}

File: packages/react-native-autoplay/android/src/main/java/com/margelo/nitro/at/g4rb4g3/autoplay/template/MessageTemplate.kt
package com.margelo.nitro.at.g4rb4g3.autoplay.template

import androidx.car.app.CarContext
import androidx.car.app.model.MessageTemplate
import androidx.car.app.model.Template
import androidx.car.app.navigation.model.MapController
import androidx.car.app.navigation.model.MapWithContentTemplate
import com.margelo.nitro.at.g4rb4g3.autoplay.hybrid.MessageTemplateConfig
import com.margelo.nitro.at.g4rb4g3.autoplay.hybrid.NitroAction

class MessageTemplate(context: CarContext, config: MessageTemplateConfig) :
    AndroidAutoTemplate<MessageTemplateConfig>(context, config) {

    override val isRenderTemplate = false
    override val templateId: String
        get() = config.id
    override val autoDismissMs = config.autoDismissMs

    override fun parse(): Template {
        val template = MessageTemplate.Builder(Parser.parseText(config.message)).apply {
            config.title?.let { title ->
                setHeader(Parser.parseHeader(context, title, config.headerActions))
            }

            config.actions?.let { actions ->
                actions.forEach { action ->
                    addAction(Parser.parseAction(context, action))
                }
            }

            config.image?.let { image ->
                setIcon(Parser.parseImage(context, image))
            }
        }.build()

        return this.config.mapConfig?.let {
            MapWithContentTemplate.Builder().apply {
                setContentTemplate(template)
                it.mapButtons?.let { mapButtons ->
                    setMapController(
                        MapController.Builder()
                            .setMapActionStrip(Parser.parseMapActions(context, mapButtons)).build()
                    )
                }
                it.headerActions?.let { headerActions ->
                    setActionStrip(Parser.parseMapHeaderActions(context, headerActions))
                }
            }.build()
        } ?: run {
            template
        }
    }

    override fun onDidAppear() {
        config.onDidAppear?.let { it(null) }
    }

    override fun onDidDisappear() {
        config.onDidDisappear?.let { it(null) }
    }

    override fun onPopped() {
        config.onPopped?.let { it() }
        templates.remove(templateId)
    }

    override fun onWillAppear() {
        config.onWillAppear?.let { it(null) }
    }

    override fun onWillDisappear() {
        config.onWillDisappear?.let { it(null) }
    }

    override fun setTemplateHeaderActions(headerActions: Array<NitroAction>?) {
        config = config.copy(headerActions = headerActions)
        super.applyConfigUpdate()
    }

    companion object {
        const val TAG = "MessageTemplate"
    }
}